server-worker/src/main/scala/com/aliyun/emr/rss/service/deploy/worker/LocalStorageManager.scala

```diff
@@ -292,14 +294,18 @@ private[worker] final class LocalStorageManager(
 
   deviceMonitor.startCheck()
 
-  def workingDirsSnapshot(mountPoint: String): util.ArrayList[File] = {
-    if (mountPoint != StorageInfo.UNKNOWN_DISK) {
-      new util.ArrayList[File](diskInfos.get(mountPoint)
-        .dirInfos.filter(workingDirs.contains(_)).toList.asJava)
-    } else {
-      logDebug("mount point is invalid")
-      workingDirsSnapshot()
-    }
+  def getUsableWorkingDirs(mountPoint: String): util.ArrayList[File] = {
+    new util.ArrayList(
+      if (mountPoint != StorageInfo.UNKNOWN_DISK
+        && diskInfos.get(mountPoint).usableSpace > diskMinimumReserveSize) {
+        new util.ArrayList[File](diskInfos.get(mountPoint)
+          .dirInfos.filter(workingDirs.contains(_)).toList.asJava)
+      } else {
+        logDebug(s"mount point $mountPoint is invalid or run out of space")
+        workingDirsSnapshot()
+      }.asScala.filter { dir =>
+        workingDirDiskInfos.get(dir.getAbsolutePath).usableSpace > diskMinimumReserveSize
```

Comment:
Should filter after workingDirsSnapshot() since the other branch has already filtered out

Reply:
OK, I moved the filter.